smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java

```diff
@@ -0,0 +1,349 @@
+ * and deserialization of a Smithy trace file.
+ */
+public final class TraceFile implements ToNode, ToSmithyBuilder<TraceFile> {
+    public static final String SMITHY_TRACE_TEXT = "smithyTrace";
+    public static final String ARTIFACT_TEXT = "artifact";
+    public static final String DEFINITIONS_TEXT = "definitions";
+    public static final String SHAPES_TEXT = "shapes";
+    public static final String SMITHY_TRACE_VERSION = "1.0";
+
+    private String smithyTrace;
+    private ArtifactMetadata artifactMetadata;
+    private ArtifactDefinitions artifactDefinitions; //Optional
+    private Map<ShapeId, List<ShapeLink>> shapes;
+    private SourceLocation sl = new SourceLocation("");
+
+    private TraceFile(Builder builder) {
+        smithyTrace = SmithyBuilder.requiredState(SMITHY_TRACE_TEXT, builder.smithyTrace);
+        artifactMetadata = SmithyBuilder.requiredState(ARTIFACT_TEXT, builder.artifactMetadata);
+        if (builder.shapes.isEmpty()) {
+            throw new IllegalStateException("TraceFile's shapes field must not be empty to build it.");
+        }
+        shapes = MapUtils.copyOf(builder.shapes);
+        artifactDefinitions = builder.artifactDefinitions;
+    }
+
+    /**
+     * Converts ObjectNode into TraceFile.
+     *
+     * @param value an ObjectNode that represents the entire trace file.
+     */
+    public static TraceFile createFromNode(Node value) {
+        ObjectNode node = value.expectObjectNode();
+        Builder builder = builder()
+                .smithyTrace(node.expectStringMember(SMITHY_TRACE_TEXT).getValue())
+                .artifact(ArtifactMetadata.createFromNode(node.expectObjectMember(ARTIFACT_TEXT)));
+
+        //parse shapes
+        Map<StringNode, Node> shapeMap = node.expectObjectMember(SHAPES_TEXT).getMembers();
+        for (Map.Entry<StringNode, Node> entry : shapeMap.entrySet()) {
+            for (Node linkNode : (entry.getValue().expectArrayNode()).getElements()) {
+                builder.addShapeLink(entry.getKey().getValue(), ShapeLink.createFromNode(linkNode));
+            }
+        }
+
+        //parse definitions
+        if (node.containsMember(DEFINITIONS_TEXT)) {
+            builder.definitions(ArtifactDefinitions.createFromNode(node.expectObjectMember(DEFINITIONS_TEXT)));
+        }
+
+        return builder.build();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Converts TraceFile instance variables into an
+     * ObjectNode.
+     *
+     * @return ObjectNode representation of a TraceFile.
+     */
+    @Override
+    public ObjectNode toNode() {
+        //constructing shapes ObjectNode map
+        ObjectNode.Builder shapesBuilder = ObjectNode.objectNodeBuilder();
+        for (Map.Entry<ShapeId, List<ShapeLink>> entry : shapes.entrySet()) {
+            shapesBuilder.withMember(entry.getKey().toString(),
+                    entry.getValue().stream().map(ShapeLink::toNode).collect(ArrayNode.collect()));
```

Comment:
Minor: This value generation is really long, consider storing it in a variable to make it easier to read.